### PR TITLE
Extract country polygon via get() function

### LINKED ISSUE #4: [GPKG] Add function to process gpkg using ArchGDAL package

### PR DESCRIPTION
**Pull request by zerefwayne**
Closing #4 

```julia
julia> Geography.get("IND")
Geometry: MULTIPOLYGON (((93.78772736 6.85264015,93.7884903  ... 49)))
```

### CODE REVIEW

File: src/polygon.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
+"""
+    getlevel(dataset::ArchGDAL.AbstractDataset, reqlevel::Int) 
+Get layer of the desired level from the dataset
+"""
+function getlevel(dataset::ArchGDAL.AbstractDataset, reqlevel::Int)
```
Comment (juliohm) [inline src/polygon.jl:5]: I would remove the type annotation, or at least write it as dataset::AbstractDataset instead.

File: src/polygon.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    # number of layers in the dataset
+    layers = ArchGDAL.nlayer(dataset)
```
Comment (juliohm) [inline src/polygon.jl]: I would call this variable `nlayers` to emphasize it is a count.

File: src/polygon.jl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    for i = 0:layers-1
+        layer = ArchGDAL.getlayer(dataset, i)
+        layername = ArchGDAL.getname(layer)
+        layerlevel = split(layername, "_")[end]
+        if layerlevel == string(reqlevel)
+            return layer
+        end
+    end
```
Comment (juliohm) [inline src/polygon.jl]: Why loop if you are only returning layer i?

File: src/polygon.jl [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    throw(ArgumentError("valid levels for the given dataset are 0...$(layers-1)"))
+end
+
+"""
+    isgpkg(file)::Bool
+Returns true for .gpkg files
+"""
+function isgpkg(file)::Bool
```
Comment (juliohm) [inline src/polygon.jl]: I would remove the ::Bool annotation. These are good assertions, but the code looks inconsistent with some function having annotations and some others without annotations.

File: src/polygon.jl [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    match(r"\b[A-Z]{3}\b", code) !== nothing
+end
+
+"""
+    get(code)
+Returns the MULTIPOLYGON data for the requested country.\n
+Input: ISO3 Country Code
+"""
+function get(code)
+
```
Comment (juliohm) [inline src/polygon.jl]: Extra white line.

File: src/polygon.jl [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    # only uppercase country codes are accepted
+    isvalidcode(code) || throw(ArgumentError("please provide standard ISO 3 country codes"))
+
+    dataid = "GADM_$code"
+
+    resource_data_path = download(dataid)
```
Comment (juliohm) [inline src/polygon.jl]: We shouldn't be calling download everytime a call to `get` is performed. Instead we should assume that the dataset is available. If it is not available, then we ask the user if he/she wants to download the data with a YES/NO prompt.

File: src/polygon.jl [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    dataset_gpkg = filter(isgpkg, files)[1]
+
+    dataset = ArchGDAL.read(dataset_gpkg)
+
+    # get country level 0 layer
+    country_layer = getlevel(dataset, 0)
+
+    geom = ArchGDAL.getfeature(country_layer, 1) do feature
+        ArchGDAL.getgeom(feature)        
+    end
```
Comment (juliohm) [inline src/polygon.jl]: These lines of code belong to the same idea and should live in a separate function. The function takes a dataset and returns the geometry, no matter the intermediate steps.

File: src/polygon.jl [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    !isnothing(geom) ? geom : @error "failed to extract the requested layer"
+
```
Comment (juliohm) [inline src/polygon.jl:84]: Extra white line. Please pay more attention to extra lines in next PRs :+1:

File: test/polygon.jl [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# DataDeps always downloads the data if it doesn't exist
+ENV["DATADEPS_ALWAYS_ACCEPT"] = true
```
Comment (juliohm) [inline test/polygon.jl]: This variable should be set only once in the top runtests.jl file.

File: test/polygon.jl [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    dataset_gpkg = Geography.filter(Geography.isgpkg, files)[1]
+    dataset = ArchGDAL.read(dataset_gpkg)
+
+    # incorrect level
+    @test_throws ArgumentError Geography.getlevel(dataset, 10)
+
+    # correct level
+    feature_layer = Geography.getlevel(dataset, 0)
+    @test typeof(feature_layer) === ArchGDAL.IFeatureLayer
+
```
Comment (juliohm) [inline test/polygon.jl]: Extra white line.

**Review by juliohm — CHANGES_REQUESTED** [review]
Awesome to see this getting alive! Attached are some review comments. It looks like some tests failed on TravisCI.

File: src/polygon.jl (hunk 3 above)
Comment (zerefwayne) [reply src/polygon.jl]: There are multiple layers with names like `gadm_ind36_0`, `gadm_ind36_1`, and they are not necessarily ordered. Hence, I need to loop over all layers to see which one is required. Also, it is not necessary that `gadm_ind36_0` will be the first element. Therefore looping seems to be the simplest solution.

File: src/polygon.jl (hunk 6 above)
Comment (zerefwayne) [reply src/polygon.jl]: `download(dataid)` is our function that is being called from `download.jl`. 

DataDeps library does the same job as you described. It checks first if the dataset is available, if yes, it simply returns the path of the dataset. Data will not be downloaded every time. Also, DataDeps doesn't provide any specific function to check if dataset exists or not.

File: src/polygon.jl (hunk 3 above)
Comment (juliohm) [reply src/polygon.jl]: Yes, let's get it started. I think we will need to brainstorm a more user friendly API after we get this one merged :+1:

File: src/polygon.jl (hunk 7 above)
Comment (zerefwayne) [reply src/polygon.jl]: I had this idea while writing test cases. :+1: I'll refactor this.

File: src/polygon.jl (hunk 6 above)
Comment (juliohm) [reply src/polygon.jl]: What we need to do is to properly create a registry of the datasets downloaded. Right now we are relying too much on what DataDeps does. I would make sure that the download function in this package post-processes the data, and places it in a format that is easy to check later. Your post_fetch_method is currently just unpack, but I think we could unpack and do other things like rename the folder to the country code, and add an entry to a text file somewhere saying that the dataset is available. Let's first make sure that the other issues are fixed, and that the tests are passing. I will then try to install the package and experiment with it.

Comment (juliohm) [reply src/polygon.jl]: We can discuss this more over our call soon.

**Review by juliohm — APPROVED** [review]

**Pull request merged**